lib/enum.ml

```diff
@@ -0,0 +1,204 @@
+    | {ptype_kind = Ptype_variant constructors; _} when not (Utils.constructors_are_bare constructors)
+      ->
+        Raise.Enum.unhandled_type_kind ~loc "variant with arguments"
+    | {ptype_kind = Ptype_variant constructors; ptype_name = {txt = type_name; _}; _}
+      ->
+        from_enummable_variant ~loc ~type_name ~constructors
+
+  (** By giving this to the Deriving.Generator.make_noarg function below, ppxlib
+   *  will apply the function the parameters:
+   *  ~loc: Information about the current location in the code base (file, lineno etc)
+   *  ~path: The current file path?
+   *  rec_flag: ???
+   *  type_declarations: A list of the type declarations at the point of call
+   *)
+  let from_type_decl ~loc ~path:_ (_rec_flag, type_declarations) =
+    List.flatten @@ List.map (from_type_declaration ~loc) type_declarations
+end
+
+module Sig = struct
+  let to_string_function_val ~loc ~type_name =
+    let function_name = Utils.to_string_function_name ~enum_name:type_name in
+    let type_lident = {txt = Lident type_name; loc} in
+    let lhs_type = Ast_builder.Default.ptyp_constr ~loc type_lident [] in
+    let type_ =
+      Ast_builder.Default.ptyp_arrow
+        ~loc
+        Nolabel
+        lhs_type
+        [%type: string]
+    in
+    let value_description =
+      Ast_builder.Default.value_description
+      ~loc
+      ~name:{txt = function_name; loc}
+      ~type_
+      ~prim:[]
+    in
+    Ast_builder.Default.psig_value ~loc value_description
+
+  let from_string_function_val ~loc ~type_name =
+    let function_name = Utils.from_string_function_name ~enum_name:type_name in
+    let type_lident = {txt = Lident type_name; loc} in
+    let rhs_type = Ast_builder.Default.ptyp_constr ~loc type_lident [] in
+    let type_ =
+      Ast_builder.Default.ptyp_arrow
+        ~loc
+        Nolabel
+        [%type: string]
+        rhs_type
+    in
+    let value_description =
+      Ast_builder.Default.value_description
+      ~loc
+      ~name:{txt = function_name; loc}
+      ~type_
+      ~prim:[]
+    in
+    Ast_builder.Default.psig_value ~loc value_description
+
+  let from_enummable_variant ~loc ~type_ =
+    let {ptype_name = {txt = type_name; _}; _} = type_ in
+    [ to_string_function_val ~loc ~type_name
+    ; from_string_function_val ~loc ~type_name
+    ]
+
+  let from_variant ~loc ~type_ =
+    match type_ with
+    | {ptype_kind = Ptype_variant constructors; _} when not (Utils.constructors_are_bare constructors)
+      ->
+        Raise.Enum.unhandled_type_kind ~loc "variant with arguments"
+    | {ptype_params; _} when ((List.length ptype_params) > 0)
+      ->
+        Raise.Enum.unhandled_type_kind ~loc "parametrized variant"
+    | _
+      ->
+        from_enummable_variant ~loc ~type_
+
+  let from_type_declaration ~loc type_ =
+    let {ptype_kind; _} = type_ in
+    match ptype_kind with
+    | Ptype_variant _ -> from_variant ~loc ~type_
+    | Ptype_record _ -> Raise.Enum.unhandled_type_kind ~loc "record"
```

Comment:
You should probably use `ptype_loc` here instead of loc. I don't think it really matters for `ppx_enum` here since you probably deal with:
```
type t1 = ...
and type 2 = ...
```
but I think it's a good practice when writing ppx-es!

Reply:
Fixed here and in the str version